pkg/kstatus/watcher/watcher_dynamic.go

```diff
@@ -0,0 +1,825 @@
+
+		rs, err := w.readStatusFromObject(ctx, obj)
+		if err != nil {
+			handleError(eventCh, fmt.Errorf("failed to compute object status: %s: %w", id, err))
+			return
+		}
+
+		if object.IsCRD(obj) {
+			w.onCRDAdd(ctx, obj)
+		}
+
+		if isUnschedulable(rs) {
+			// schedule delayed failure & save the cancelFunc to cancel if there's an update
+			w.delayedEventCanceller.Add(id, w.scheduleFailure(ctx, eventCh, id))
+		}
+
+		eventCh <- event.Event{
+			Type:     event.ResourceUpdateEvent,
+			Resource: rs,
+		}
+	}
+
+	handler.UpdateFunc = func(_, iobj interface{}) {
+		obj, ok := iobj.(*unstructured.Unstructured)
+		if !ok {
+			panic(fmt.Sprintf("UpdateFunc received unexpected object type %T", iobj))
+		}
+		id := object.UnstructuredToObjMetadata(obj)
+		if w.ObjectFilter.Filter(obj) {
+			klog.V(7).Infof("UpdateFunc: Watch Event Skipped: %s", id)
+			return
+		}
+		klog.V(5).Infof("UpdateFunc: Computing status for object: %s", id)
+
+		// cancel any delayed events for this object
+		w.delayedEventCanceller.Cancel(id)
+
+		rs, err := w.readStatusFromObject(ctx, obj)
+		if err != nil {
+			handleError(eventCh, fmt.Errorf("failed to compute object status: %s: %w", id, err))
+			return
+		}
+
+		if object.IsCRD(obj) {
+			w.onCRDUpdate(ctx, obj)
+		}
+
+		if isUnschedulable(rs) {
+			// schedule delayed failure & save the cancelFunc to cancel if there's an update
+			w.delayedEventCanceller.Add(id, w.scheduleFailure(ctx, eventCh, id))
+		}
+
+		eventCh <- event.Event{
+			Type:     event.ResourceUpdateEvent,
+			Resource: rs,
+		}
+	}
+
+	handler.DeleteFunc = func(iobj interface{}) {
+		obj, ok := iobj.(*unstructured.Unstructured)
+		if !ok {
+			panic(fmt.Sprintf("DeleteFunc received unexpected object type %T", iobj))
+		}
+		id := object.UnstructuredToObjMetadata(obj)
+		if w.ObjectFilter.Filter(obj) {
+			klog.V(7).Infof("DeleteFunc: Watch Event Skipped: %s", id)
+			return
+		}
+		klog.V(5).Infof("DeleteFunc: Computing status for object: %s", id)
+
+		// Cancel any delayed events for this object
+		w.delayedEventCanceller.Cancel(id)
+
+		if object.IsCRD(obj) {
+			w.onCRDDelete(obj)
+		}
+
+		eventCh <- event.Event{
+			Type:     event.ResourceUpdateEvent,
+			Resource: deletedStatus(id),
+		}
+	}
+
+	return handler
+}
+
+// onCRDAdd handles creating a new informer to watch the new resource type.
+func (w *ObjectStatusInformer) onCRDAdd(ctx context.Context, obj *unstructured.Unstructured) {
+	groupKind, found := object.GetCRDGroupKind(obj)
+	if !found {
+		id := object.UnstructuredToObjMetadata(obj)
+		klog.Warningf("Invalid CRD added: missing group and/or kind: %v", id)
+		// Don't return an error, because this should not inturrupt the task queue.
+		// TODO: Allow non-fatal errors to be reported using a specific error type.
+		return
+	}
+	klog.V(3).Infof("CRD added for %s", groupKind)
+
+	klog.V(3).Info("Resetting RESTMapper")
+	// Reset mapper to invalidate cache.
+	meta.MaybeResetRESTMapper(w.Mapper)
```

Comment:
Do we need to do this here? We could wait until the CRD becomes established, but I'm not sure if we are guaranteed to get that update callback if the CRD becomes established immediately.

Reply:
AFAIK, the informers de-dupe missed events automatically. So if you miss an add and an update, it just gives you an add. But even if we check the status and only reset if it's reconciled, there's still some time between when it became reconciled and when we get the event that it was reconciled that the CRD is registered but our mapper cache says it's not. 

Since reset itself is cheap (it's only expensive when you make the next mapping request), I opted to reset in both add and update. If we get no other events between the add and the update, and no other async code does a lookup, the mapper will still be reset when we get the update event. But if there IS an async mapping call between, it will fetch the latest mappings, rather than remain potentially out of date.

Also, FWIW, the kernel team is working on some enhancements upstream to make resetting cheaper, so that it can hit an endpoint to determine which cached CRDs to invalidate.